cpp/src/arrow/dataset/file_parquet.cc

```diff
@@ -36,38 +37,91 @@ namespace dataset {
-        reader_(reader),
-        options_(std::move(options)),
-        context_(std::move(context)) {}
+      : reader_(reader), options_(std::move(options)), context_(std::move(context)) {
+    size_t i = 0;
+    row_groups_ = MakeFunctionIterator([row_groups, i](int** out) mutable {
+      if (i < row_groups.size()) {
+        *out = &row_groups[i++];
+      } else {
+        *out = nullptr;
+      }
+      return Status::OK();
+    });
+  }
 
   RecordBatchIterator Scan() {
+    auto options = options_;
+    auto reader = reader_;
+    auto batches_it = MakeMaybeMapIterator(
+        [reader, options](int* row_group, RecordBatchIterator* out) {
+          return ScanRowGroup(*row_group, reader, options, out);
+        },
+        std::move(row_groups_));
+    return MakeFlattenIterator(std::move(batches_it));
+  }
+
+ private:
+  static Status ScanRowGroup(int row_group,
+                             const std::shared_ptr<parquet::arrow::FileReader>& reader,
+                             const std::shared_ptr<ScanOptions>& options,
+                             RecordBatchIterator* out) {
+    auto metadata = reader->parquet_reader()->metadata()->RowGroup(row_group);
+    std::vector<int> columns_projection;
+    if (!InferColumnProjection(*metadata, options, &columns_projection).ok()) {
+      // fall back to trivial projection
+      columns_projection = internal::Iota(metadata->num_columns());
+    }
+
     // The construction of parquet's RecordBatchReader is deferred here to
     // control the memory usage of consumers who materialize all ScanTasks
     // before dispatching them, e.g. for scheduling purposes.
     //
     // Thus the memory incurred by the RecordBatchReader is allocated when
     // Scan is called.
     std::unique_ptr<RecordBatchReader> record_batch_reader;
-    auto status = reader_->GetRecordBatchReader(row_groups_, columns_projection_,
-                                                &record_batch_reader);
-    // Propagate the previous error as an error iterator.
-    if (!status.ok()) {
-      return MakeErrorIterator<std::shared_ptr<RecordBatch>>(std::move(status));
+    RETURN_NOT_OK(reader->GetRecordBatchReader({row_group}, columns_projection,
+                                               &record_batch_reader));
+    *out = MakePointerIterator(std::move(record_batch_reader));
+    return Status::OK();
+  }
+
+  // Compute the column projection out of an optional arrow::Schema
+  static Status InferColumnProjection(const parquet::RowGroupMetaData& metadata,
+                                      const std::shared_ptr<ScanOptions>& options,
+                                      std::vector<int>* out) {
+    if (options->projector == nullptr) {
+      return Status::Invalid("no projector available");
     }
 
-    return MakePointerIterator(std::move(record_batch_reader));
+    // union fields in expression and fields in projector
+    auto projector_schema = options->projector->schema();
```

Comment:
I don't understand the logic of mixing the projector's schema and the option's schema. Why can't it just use the option schema?